src/main/java/tests/ShopServiceTest.java

```diff
@@ -0,0 +1,110 @@
+package tests;
+
+import core.basesyntax.*;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class ShopServiceTest {
+
+    private ShopService shopService;
+    private List<FruitTransaction> transactions;
+
+    @BeforeEach
+    public void setUp() {
+        Map<Operation, OperationHandler> operationHandlers = new HashMap<>();
+        operationHandlers.put(Operation.BALANCE, new BalanceOperation());
+        operationHandlers.put(Operation.PURCHASE, new PurchaseOperation());
+        operationHandlers.put(Operation.RETURN, new ReturnOperation());
+        operationHandlers.put(Operation.SUPPLY, new SupplyOperation());
+        OperationStrategy operationStrategy = new OperationStrategyImpl(operationHandlers);
+        shopService = new ShopServiceImpl(operationStrategy);
+        transactions = new ArrayList<>();
+    }
+
+    @Test
+    public void test_balance_operation() {
+        FruitTransaction balance = new FruitTransaction();
+        balance.setOperation(Operation.BALANCE);
+        balance.setFruit("banana");
+        balance.setQuantity(100);
+        transactions.add(balance);
+
+        shopService.process(transactions);
+        Map<String, Integer> storage = shopService.getStorage();
+
+        assertEquals(100, (int) storage.get("banana"));
+    }
+
+    @Test
+    public void test_supply_operation() {
+        FruitTransaction balance = new FruitTransaction();
+        balance.setOperation(Operation.BALANCE);
+        balance.setFruit("banana");
+        balance.setQuantity(100);
+        transactions.add(balance);
+
+        FruitTransaction supply = new FruitTransaction();
+        supply.setOperation(Operation.SUPPLY);
+        supply.setFruit("banana");
+        supply.setQuantity(50);
+        transactions.add(supply);
+
+        shopService.process(transactions);
+        Map<String, Integer> storage = shopService.getStorage();
+
+        assertEquals(150, (int) storage.get("banana"));
+    }
+
+    @Test
+    public void test_purchase_operation() {
+        FruitTransaction balance = new FruitTransaction();
+        balance.setOperation(Operation.BALANCE);
+        balance.setFruit("banana");
+        balance.setQuantity(100);
+        transactions.add(balance);
+
+        FruitTransaction purchase = new FruitTransaction();
+        purchase.setOperation(Operation.PURCHASE);
+        purchase.setFruit("banana");
+        purchase.setQuantity(30);
+        transactions.add(purchase);
+
+        shopService.process(transactions);
+        Map<String, Integer> storage = shopService.getStorage();
+
+        assertEquals(70, (int) storage.get("banana"));
```

Comment:
After a purchase operation, it's important to check not only that the quantity has decreased but also that it's not possible to purchase more than the available quantity. Consider adding a test case for a scenario where a purchase operation attempts to buy more than the available stock.